src/librustc_middle/mir/mod.rs

```diff
@@ -155,6 +155,9 @@ pub struct Body<'tcx> {
     /// A span representing this MIR, for error reporting.
     pub span: Span,
 
+    /// Unevaluated consts to evaluate them regardless of being optimized out
+    pub uneval_consts: Vec<Constant<'tcx>>,
```

Comment:
Is "unevaluated" really accurate here? Aren't these just all the constants used in this function? (That is a scheme we considered in https://github.com/rust-lang/rust/issues/67191, and then maybe making consts inside the body refer to this `Vec` by index.)

Reply:
Oh I see, we collect all `ConstKind::Unevaluated` into this.
Is there a good way to say in terms of surface Rust which constants these are? Like, does it include all uses of `const` items, but *not* include literals as in `1`?

Reply:
@RalfJung from what you've said I ended thinking it would be great if we can document each of those `ConstKind` things. To be honest I can't answer 100% clear about each of the different constants, I know that Unevaluated are the ones left unevaluated until post monomorphization, the rest are already evaluated so not related to this issue.

Anyway cc @oli-obk, would be great if we can document this data structure with examples and also other data structures :).

Reply:
That entirely depends on the point where you are in the optimization pipeline. Ideally we'd only have constants that are too generic to be evaluated, but at this point we include all non-literal constants.

Reply:
"this point" is not anywhere in the pipeline, this is the data structure that is used throughout the pipeline. Optimizations don't remove entries from here even when they optimize MIR to contain evaluated consts, right?

Maybe `referred_consts` or so would be a better name? There's no type for the `(DefId, SubstsRef<'tcx>, Option<Promoted>)` triple that we could use here to ensure that this only stores unevaluated consts, is there?

Reply:
`so_far_unevaluable_constants` is what it is supposed to be. We can remove any constants at any point where they are evaluable.

I don't want to use that triple, as ideally some day in the future we can do the MIR 2.0 change of making this an `IndexVec` and referring to all constants indirectly via index instead of storing them inline in the MIR datastructures.

Reply:
> `so_far_unevaluable_constants`

More like `unevaluated`? Still, that's it very informative.

What about `required_const`, with doc comment  like "constants that are required to evaluated successfully for this MIR to be well-formed".

> I don't want to use that triple,

Fair.

Reply:
> What about required_consts, with a doc comment like "constants that are required to evaluate successfully for this MIR to be well-formed".

I like that

Reply:
Done